Tweaks to create a single-file decoder

cwoffenden opened the pull request:
The CHECK_F macros differ slightly (but eventually do the same thing) so the guards allow all the decoder files to be 'amalgamated' into a single file.

Older GCC needs to fallback on the old-style pragma optimisation flags (the alternative path was verified with GCC4).

Cyan4973 commented:
That's an interesting suggestion @cwoffenden .

So amalgamating the decoder source code is a thing ?

Maybe we need to provide a script to generate such amalgamation in an "official" way ?
Also, maybe we need to add a test to ensure it still works in the future ?

While I'm sure your patch fixes the issue right now, any future change might break it again, and we won't notice it without a CI test.

cwoffenden commented:
We've been using the decompressor this way for a while, dropping it in as a single file with a bunch of config and optimisation options rolled in. For folks new to Zstd, having a single file with no dependencies makes for a very quick introduction.

I realise as it stands my changes aren't very useful (maybe apart from the older GCC changes) so I'm currently writing a simple amalgamation script, then I'll add an example and test. We'd previously been generating the single file with a native tool in our fork but a bash script will do the trick.

cwoffenden commented:
I wrote a Bash script that will combine the decompressor sources (`combine.sh`) and a simple test. The resulting single source file is 493kB.

The simple test here just checks that it builds and the decompression is correct. I'll add a further test with the streaming decompressor (I already dropped it into an existing project and the generated source works just fine, tested on Windows, Mac, iOS and Emscripten, but I'll create another small standalone test).

I guess the next step is to automate the script and the test. The script doesn't do anything fancy, it just inlines any includes, skipping previously seen ones.

cwoffenden commented:
I'm happy with the generator script and will call that done but I'm not happy with the tests/examples. I'm going to remove the Buck files and add the tests either as a makefile or shell script (if you have a preference do say; I spent far too long fighting with Buck and GTest for something so simple).

I expect to be finished tomorrow.

cwoffenden commented:
Now I think it's worth looking at...! I added `build.sh` which isn't elegant but creates the library, builds and tests the decompression, builds the Emscripten example if possible, reporting what passed or failed (also in the exit code), and cleans up after itself.

A compiled version of the Emscripten test [is on-line here](https://wip.numfum.com/2019-08-27/index.html) (I could add this to the sample but it's easy enough to build for those interested).

I added in the readme some reasoning for amalgamating the source: in short, one file with an optional header is all it takes to give Zstd decompression support, with no further integration work on the developer's part.

cwoffenden commented:
Thanks! If you could add it that'd save me from poking around to find _how_ to add it (or just tell me which of the config files and I'll take a look).

In the meantime I fixed the test script to not fail if the Emscripten compiler isn't present.

cwoffenden commented on contrib/declib/combine.sh:
```diff
@@ -0,0 +1,111 @@
+#!/bin/bash
```
It'll need the regex lines rewriting and the `sh` implementation will need to support locals (due to the recursion) but sure, no problem. I'll need to see what I can find to test on (the Mac's `sh` is `bash`).

cwoffenden replied:
Added PR #1750. It's much slower but should work everywhere. Tested with `dash` which I _think_ is the `sh` variant from modern BSDs.
